### ISSUE TITLE
Multiprocessing thoughts

### ISSUE THREAD

**Issue by guynir42** [c3-time-domain/SeeChange#14]
The main problem is that of running multiple instances to process different datasets in parallel. 

Here are some conclusions we have made from our chat (Rob & Guy):
- You can run each CCD separately except for places where we download common data like calibration references. 
- In that case you may want to lock, lazy load the data into local storage, and then let other threads grab the new data and lazy load it for themselves. 
- We need a locking mechanism but it is not certain we need any more communication between threads/processes, so we may be able to get away without using MPI. 
- Ideally the parallelization will work the same with a single core, a single local machine with multiple cores, or a task divided into a cluster's nodes by a queue mechanism. 
- Partially complete phases will have to be re-processed from scratch, and only data from completed phases (such that was provenance-stamped and save to DB) can be used as input to the next phase. This prevents picking up partial or wrong data on a local machine. 

Please add any more thoughts or ideas to the discussion.

**Comment by rknop** [comment]
I think based on our discussion that we should not try to use MPI to coordinate between chips.  As such, most of the code won't need MPI.

We might want to use MPI *within* a chip sometimes (e.g. when fitting a variable sky using the bijaoui sky algorithm), and eventually we definitely want to use Cuda for stuff, but I don't think we should bake MPI into the structure of the whole thing.

I think the locking can be handled with the database and proper use of transactions and table locking.  (So, you want to get a conceptual code lock.  First, lock the postgres table that has the lock, read to see if there already is a lock, if so, unlock the table, sleep for a second (perhaps?) and loop back.  If there isn't a lock, create a lock, unlock the table, and move on.  The table locking prevents the race condition of two processes at the same time reading the database and both not seeing a lock before one of them is able to put the lock in.

**Comment by guynir42** [comment]
Two things:
1) We could just as well lock using a new file. I am not sure why that would be less efficient or robust than putting a lock on the DB. That way the "lock file" exists on the local file system, the same place you are making the newly fetched data. If you lock a DB table you also need to keep track of which node you are locking for, and the nodes will be able to lock each other out (which is not what you want). I need to read about how to do this in such a way that is race-condition free. 

2) I have been thinking about the nested-multithreading of longer tasks like sky subtractions. 
It seems to me that on a pipeline level it gives no benefit to parallelize any of the sub-tasks. You will most likely be running all your cores doing multiple images at once. If you split an internal task to those cores, you'd be blocking the other cores from running. 
If you run these tasks in standalone mode, it may be nice to have it return x4 times faster, but I'm not sure that warrants the additional coding. And if so, maybe a light-weight "threading" implementation would be a better idea, just for a specific long running task.

**Comment by rknop** [comment]
So, there's two different lock cases here.

In the case of fetching a file to the local filesystem, then, yes, you could use a local lock file, and I think there is a standard way of doing that race-condition free (though I don't remember what that is -- I'm pretty sure I've used this before, though perhaps from C).

However, we may still need database level locking for updating stuff.   The pipeline may potentially be running on multiple systems at once.  The only thing that's guaranteed to be shared between different systems working on the same database is the global archive (which for lensgrinder is at NERSC) and the database, so the for those kinds of locks, the database is the appropriate place to do it.   (Another way is to have a dedicated lock server out there.  Effectively, this is something the conductor does, although not explcitly.  When a process wants to start working on an exposure, it doesn't read the knownexposures table directly, it asks the conductor what it should do, so the conductor is the only process updating that table.)

Re: multithreading stuff, one use case is during development; you want your code to run faster so you can iterate faster.  My eyes would have bled when I was trying to do fast tests and iteration on bijaouisky if I hadn't set it up to be able to use 20 processes at once.  In any event, though, I don't think we should *disallow* multithreading using either MPI or python multiprocessing.  Or, anything else -- openmp in C, say.  (Some of the command line tools that we  or whatever inside individual steps.  We do have to think about not overfilling cores, so there needs to be a way of making sure that too many processes aren't run at once, thereby overloading the cores.

One reason not to use MPI is that I *think* that if you want to do things reasonably, the outside level of your code has to be MPI-aware to avoid doing things wrong.  If I just wanted to implement multiprocessing in (say) bijaouisky, while what I have right now uses MPI -- which was reasonable since the whole pipeline used MPI -- it would probably make more sense to develop something using python multiprocessing.  That way, no code other than the one bit of code I'm parallelizing never needs to know about it, *except* for needing to have some config parameter somewhere that tells processes how many CPUs they can use.

Eventually we're going to want to be able to write Cuda (or other GPU code) to do parts of things, which will of course add a whole 'nuther layer of multiprocessing.

**Comment by guynir42** [comment]
I generally agree. The specific code for running slower parts of the pipeline can be optionally multi-theaded if that makes developing faster. 

I am not sure why we would need a database level lock, if each process only works on its own images.

**Comment by guynir42** [comment]
May be related: https://www.gorgias.com/blog/prevent-idle-in-transaction-engineering
